pallets/aggregate/src/lib.rs

```diff
@@ -0,0 +1,608 @@
+            id: u64,
+        ) -> DispatchResultWithPostInfo {
+            use frame_support::traits::DefensiveSaturating;
+            let origin = ensure_signed(origin)?;
+            let (root, size) = Domains::<T>::try_mutate(domain_id, |domain| {
+                let domain = domain.as_mut().ok_or_else(|| {
+                    dispatch_post_error(
+                        T::WeightInfo::aggregate_on_invalid_domain(),
+                        Error::<T>::UnknownDomainId,
+                    )
+                })?;
+                let aggregation = domain.should_publish.remove(&id).ok_or_else(|| {
+                    dispatch_post_error(
+                        T::WeightInfo::aggregate_on_invalid_id(),
+                        Error::<T>::InvalidAggregationId,
+                    )
+                })?;
+
+                let root = aggregation.compute();
+                let size = aggregation.size;
+                Published::<T>::mutate(|published: &mut _| {
+                    published.push((domain_id, aggregation))
+                });
+
+                if let Some((_, published)) = Published::<T>::get().last() {
+                    for s in published.statements.iter() {
+                        let account = &s.account;
+                        let remain = T::Hold::transfer_on_hold(
+                            &HoldReason::Aggregation.into(),
+                            account,
+                            &origin,
+                            s.reserve,
+                            Precision::BestEffort,
+                            Restriction::Free,
+                            Fortitude::Polite,
+                        )
+                        .expect("Call user should exists. qed")
+                        .defensive_saturating_sub(s.reserve);
+                        if remain > 0_u32.into() {
+                            log::warn!(
+                                "Cannot refund all founds from {account:?} to {origin:?}: missed {remain:?}"
+                            )
+                        }
+                    }
+                }
+
+                Result::<_, DispatchErrorWithPostInfo>::Ok((root, size))
+            })?;
+            Self::deposit_event(Event::NewAggregationReceipt {
+                domain_id,
+                aggregation_id: id,
+                receipt: root,
+            });
+            Ok(Some(T::WeightInfo::aggregate(size as u32)).into())
+        }
+
+        #[pallet::call_index(1)]
+        pub fn register_domain(
+            origin: OriginFor<T>,
+            aggregation_size: AggregationSize,
+            queue_size: Option<u32>,
+        ) -> DispatchResultWithPostInfo {
+            let id = Self::next_domain_id();
+            let owner = User::<T::AccountId>::from_origin::<T>(origin)?;
+            let queue_size = queue_size.unwrap_or(T::MaxPendingPublishQueueSize::get());
+            Self::ensure_domain_params(aggregation_size, queue_size)?;
```

Comment:
I don't like it ... but I'm open on it.... `create` is more a constructor and _should never fail_, but we can superssed it or add a new function that return an error... but we should takecare that Domain function cannot return a `DispatchError` or... is better that doesn't return this error type because `Domain` is just a data structure.

Anyway I'll try to replace the Domain type with a named tupla and the add somathing like `try_create()` to do it